packages/block-editor/src/store/effects.js

```diff
@@ -80,27 +80,33 @@ export default {
 			return;
 		}
 
-		// We can only merge blocks with similar types
-		// thus, we transform the block to merge first
-		const blockB = getBlock( state, secondBlockClientId );
+		const blockB = getBlock( state, clientIdB );
 		const blockBType = getBlockType( blockB.name );
+		const { clientId, attributeKey, offset } = getSelectionStart( state );
+		const hasSelection = clientId === clientIdA || clientId === clientIdB;
```

Comment:
>The problem is that we are just assuming that the caret is in block B when merging blocks, while it could be in either.

I guess more accurately: It could be either, _**or neither**_? Which to me sounds like a reasonable qualification to consider, since I should be able to call `mergeBlocks` on any two blocks, selected or otherwise.

Reply:
Yes, that's why I added the `hasSelection` check. If none of the blocks have selection, the caret replacement will be skipped.